unicorn/src/unicorn.cpp

```diff
@@ -0,0 +1,795 @@
+        }
+    }
+}
+
+void distributed_lock_t::lock_action_t::operator()(int rc, zookeeper::node_stat const& stat) {
+    /* If next lock in queue disappeared during request - issue childs immediately. */
+    if(rc == ZNONODE) {
+        try {
+            service->zk.childs(folder, *this);
+        }
+        catch(const zookeeper::exception& e) {
+            result.abort(e.code(), e.what());
+            state->release();
+        }
+    }
+}
+
+void distributed_lock_t::lock_action_t::operator()(int type, int zk_state, path_t path) {
+    try {
+        service->zk.childs(folder, *this);
+    }
+    catch(const zookeeper::exception& e) {
+        result.abort(e.code(), e.what());
+        state->release();
+    }
+}
+
+void
+distributed_lock_t::lock_action_t::finalize(zookeeper::value_t value) {
+    zookeeper::path_t created_path = std::move(value);
+    created_node_name = zookeeper::get_node_name(created_path);
+    if(state->set_lock_created(std::move(created_path))) {
+        try {
+            service->zk.childs(folder, *this);
+        }
+        catch(const zookeeper::exception& e) {
+            result.abort(e.code(), e.what());
+            state->release();
+        }
+    }
+}
+
+void
+distributed_lock_t::lock_action_t::abort(int rc) {
+    result.abort(rc, zookeeper::get_error_message(rc));
+}
+
+distributed_lock_t::lock_state_t::lock_state_t(unicorn_service_t* _service) :
+    service(_service),
+    lock_created(false),
+    lock_released(false),
+    discarded(false),
+    created_path(),
+    access_mutex()
+{}
+
+distributed_lock_t::lock_state_t::~lock_state_t(){
+    release();
+}
+
+void
+distributed_lock_t::lock_state_t::release() {
+    bool should_release = false;
+    {
+        std::lock_guard<std::mutex> guard(access_mutex);
+        should_release = lock_created && !lock_released;
+        lock_released = lock_released || should_release;
+    }
+    if(should_release) {
+        release_impl();
+    }
+}
+
+bool
+distributed_lock_t::lock_state_t::release_if_discarded() {
+    bool should_release = false;
+    {
+        std::lock_guard<std::mutex> guard(access_mutex);
+        if(discarded && lock_created && !lock_released) {
+            should_release = true;
+            lock_released = true;
+        }
+    }
+    if(should_release) {
+        release_impl();
+    }
+    return lock_released;
+}
+
+void
+distributed_lock_t::lock_state_t::discard() {
+    bool should_release = false;
+    {
+        std::lock_guard<std::mutex> guard(access_mutex);
+        discarded = true;
+        should_release = lock_created && !lock_released;
+        lock_released = lock_released || should_release;
+    }
+    if(should_release) {
+        release_impl();
+    }
+}
+
+bool
+distributed_lock_t::lock_state_t::set_lock_created(path_t _created_path) {
+    //Safe as there is only one call to this
+    created_path = std::move(_created_path);
+    COCAINE_LOG_DEBUG(service->log, "Created lock: %s", created_path.c_str());
```

Comment:
Log style here and elsewhere: all log messages should be started with lower character.